src/StoreMeta.cc

```diff
@@ -9,172 +9,147 @@
+        false;
 }
 
-StoreMeta *
-StoreMeta::Factory (char type, size_t len, void const *value)
+/// Whether the given raw swap meta field type can be safely ignored.
+/// \sa HonoredSwapMetaType()
+inline constexpr
+bool
+IgnoredSwapMetaType(const RawSwapMetaType type)
 {
-    if (!validType(type))
-        return NULL;
-
-    StoreMeta *result;
-
-    switch (type) {
-
-    case STORE_META_KEY:
-        result = new StoreMetaMD5;
-        break;
+    return DeprecatedSwapMetaType(type) || ReservedSwapMetaType(type);
+}
 
-    case STORE_META_URL:
-        result = new StoreMetaURL;
-        break;
+/// Whether we store the given swap meta field type (and also interpret the
+/// corresponding swap meta field when the Store loads it). Matches all
+/// SwapMetaType enum values except for the never-stored/loaded STORE_META_VOID.
+inline constexpr
+bool
+HonoredSwapMetaType(const RawSwapMetaType type)
+{
+    return 0 < type && type <= SwapMetaTypeMax && !IgnoredSwapMetaType(type);
+}
 
-    case STORE_META_STD:
-        result = new StoreMetaSTD;
-        break;
+static_assert(SwapMetaTypeMax <= std::numeric_limits<RawSwapMetaType>::max(),
+              "RawSwapMetaType fits all SwapMetaType values");
 
-    case STORE_META_STD_LFS:
-        result = new StoreMetaSTDLFS;
-        break;
+/// properly reports or rejects a problematic raw swap meta field type
+static
+void
+HandleBadRawType(const RawSwapMetaType type)
+{
+    if (ReservedSwapMetaType(type)) {
+        debugs(20, 3, "ignoring swap meta field with a reserved type: " << int(type));
+        return;
+    }
 
-    case STORE_META_OBJSIZE:
-        result = new StoreMetaObjSize;
-        break;
+    if (DeprecatedSwapMetaType(type)) {
+        debugs(20, DBG_CRITICAL, "ERROR: Ignoring swap meta field with a deprecated type: " << int(type));
+        return;
+    }
 
-    case STORE_META_VARY_HEADERS:
-        result = new StoreMetaVary;
-        break;
+    // TODO: Instead of assuming that all future swap meta types can be ignored
+    // (and some should be reported at level-0/1), future-proof this code by
+    // using a type bit to define whether to silently ignore a swap meta field
+    // with that type (or even the whole Store entry with that field).
 
-    default:
-        debugs(20, DBG_CRITICAL, "ERROR: Attempt to create unknown concrete StoreMeta");
-        return NULL;
+    if (type > SwapMetaTypeMax + 10) {
+        debugs(20, DBG_CRITICAL, "ERROR: Malformed cache storage; ignoring swap meta field with an unexpected type: " << int(type));
+        return;
     }
 
-    if (!result->validLength(len)) {
-        delete result;
-        return NULL;
+    if (type > SwapMetaTypeMax) {
+        debugs(20, 3, "ignoring swap meta field with a presumed future type: " << int(type));
+        return;
     }
 
-    result->length = len;
-    result->value = xmalloc(len);
-    memcpy(result->value, value, len);
-    return result;
+    Assure(type <= 0);
```

Comment:
Please avoid magic numbers here.

```suggestion
    Assure(type <= STORE_META_VOID);
```